src/monmove.cpp

```diff
@@ -1381,6 +1539,70 @@ bool monster::attack_at( const tripoint &p )
     return false;
 }
 
+bool monster::pickup_at( const tripoint &p, item_location &target )
+{
+    if( p != goal && has_effect( effect_looting ) ) {
+        return false;
+    }
+
+    if( target.get_item() == nullptr ) {
+        return false;
+    }
+
+    int charges_to_pick = target->charges;
+
+    if( charges_to_pick <= 0 ) {
+        return false;
+    }
+
+    int charges_picked;
+
+    //Get weight of the stack or item
+    units::mass weight = target->weight();
+    units::mass capacity = this->weight_capacity();
+
+    //Adjust volume and weight for units in a stack
+    units::mass weight_each = weight / charges_to_pick;
+    charges_picked = std::min( target->charges,  int( capacity / weight_each ) );
+
+    //Caps the amount taken to avoid taking large stacks.
+    //TODO: Allow this to be adjusted in JSON
+    charges_picked = std::min( charges_picked, 2 );
+
+    //Add item to inventory and adjust charges
+    inv.push_back( *target );
+    target->mod_charges( -charges_picked );
+    inv.back().charges = charges_picked;
+
+    //If we've taken all the stack, let's remove the item
+    if( charges_picked == charges_to_pick ) {
+        target.remove_item();
+    }
+
+    //Whether we succeed or not, let's remove the effect.
+    if( has_effect( effect_looting ) ) {
+        remove_effect( effect_looting );
+    }
+
+    //Successfully taken any?
+    if( charges_picked >= 1 ) {
+        //Notify the player
+        if( g->u.sees( *this ) && !has_flag( MF_EATS_FOOD ) ) {
+            add_msg( m_warning, _( "%1$s grabs %2$s!" ), name(), inv.back().display_name() );
+        } else {
+            return eat_from_inventory( charges_picked );
```

Comment:
Monster will eat from inventory only when player doesn't see it?

Reply:
If the monster has the flag EATS_FOOD it will eat items rather than pick them up, the if-else here will omit the “grabs” message and instead the monster will immediately eat the food.